views/charts.tmpl

```diff
@@ -0,0 +1,254 @@
+{{define "charts"}}
+<!DOCTYPE html>
+<html lang="en">
+
+{{template "html-head" "Decred Charts"}}
+<body>
+    {{template "navbar" . }}
+    <style type="text/css">
+      .dygraph-legend {
+        padding: 4px !important;
+        border: 1px solid #000;
+        border-radius: 5px;
+        pointer-events: none;
+        width: fit-content;
+      }
+      .dygraph-rangesel-bgcanvas {
+        z-index: 0 !important;
+      }
+      .dygraph-rangesel-fgcanvas {
+        z-index: 0 !important;
+      }
+      .dygraph-rangesel-zoomhandle {
+        z-index: 0 !important;
+      }
+      .dygraph-axis-label {
+        z-index: 0 !important;
+      }
+    </style>
+
+    <div class="container" style="font-size:0.8em;">
+         <br>
+        <br>
+        <br>
+        <div id="row">
+            <select id="selectBox" onchange="changeFunc();">
+                <option value="1">Ticket Price Chart</option>
+                <option value="2">Ticket Pool Size Chart</option>
+                <option value="3">Ticket Pool Value Chart</option>
+                <option value="4">Average Block Size Chart</option>
+                <option value="5">BlockChain Size Chart</option>
+                <option value="6">Number of Transactions Per Block Chart</option>
+                <option value="7">Number of Transactions Per Day Chart</option>
+                <option value="8">PoW Difficulty Chart</option>
+                <option value="9">Total Coin Supply Chart</option>
+                <option value="10">Total Fee Per Block Chart</option>
+                <option value="11">Duration Between Blocks Chart</option>
+            </select>
+        </div>
+        <br>
+        <br>
+        <div id="graphdiv" style="width:100%; height:450px; margin:0 auto;"></div>
+        <br>
+    </div>
+
+    <script type="text/javascript">
+        function safeStartTurbolinksProgress() {
+            if(!Turbolinks.supported) { return; }
+            Turbolinks.controller.adapter.progressBar.setValue(0.5);
+            Turbolinks.controller.adapter.progressBar.show();
+        }
+
+        function safeStopTurbolinksProgress() {
+            if(!Turbolinks.supported) { return; }
+            Turbolinks.controller.adapter.progressBar.setValue(1);
+            Turbolinks.controller.adapter.progressBar.hide();
+        }
+        var g = new Dygraph(
+            document.getElementById("graphdiv"),
+            "1,1\n",
+            {
+                digitsAfterDecimal: 8,
+                showRangeSelector: true,
+                legend: "follow",
+                ylabel: "Data loading...",
+                xlabel: "Data loading...",
+                title: "Charts Loading Please Wait ...",
+                labelsSeparateLines: true
+            });
+
+        var data = {{.Data}};
```

Comment:
Sending the chart data inline with in the HTML incurs a large performance penalty because it forces the browser to store it twice. Once as text inside the DOM, and again in memory once the script is parsed. On my machine I'm seeing ~2 seconds just to parse the html. During this time the browser will essentially be frozen, or in UX terminology, the end user experiences "jank".

Chrome dev tools has a great Performance tab for doing this kind of profiling:
 
<img width="1273" alt="screen shot 2018-06-15 at 4 24 38 pm" src="https://user-images.githubusercontent.com/25571523/41493097-adf473fe-70b8-11e8-99ed-1b06a586a296.png">

As first step to "dejank" this, the data instead should come from an api endpoint and fetched via xhr ( use  [$.ajax()](http://api.jquery.com/jquery.ajax/) ). That way the page can load smoothly and a loading animation can be shown to the end user while the data is requested. 

Furthermore, each dataset should have it's own endpoint, as the current payload is 25mb over the wire. And parsing that network request will again create jank. 

Another trick for transporting large datasets is to use csv format instead of json, since that way we're not repeating the column names for each row of data. So I given the nature of these datasets, we might as well go that route.

Reply:
Good feedback @gozart1.  Also you might as well make the endpoint have a param to only give you the datapoints that are useful and visible, like daily or weekly or something.  that will squish the data even more.  reload the data for zoom functions.  or just lazy load the detailed data once you have the broad framework in place if you have dynamic zooms.

Reply:
Ok had a chance to look into this PR a bit more now and I see that dygraphs expects an unparsed csv string as input. So we should definitely just return csv from the backend so it can be passed directly to the chart instead of having to transcode first.

Interactive charts backed by large datasets up to a million or so records **are** viable, but only if things are streamlined. A good example is on the homepage of [crossfilter](http://square.github.io/crossfilter/) which is a d3 based charting lib. It's quite responsive even as you interact with it to apply multi-dimensional filters on a 5 column dataset with 230k records.

Reply:
@gozart1 The current implementation loads all the data needed from the backend to the frontend, the data is pulled once and reused severally thereby reducing the effects of network latency if it occurs while one is trying to load one of the many expected graphs. A work around on this would be making ajax requests for each graph while its being plotted but this will increase the data being sent from the backend to the frontend especially if once tries to view several graphs in short period of time.